File: java-sample/src/main/java/patterns/creational/singleton/SingletonSynchronized.java
package patterns.creational.singleton;

// Double-Locking
public class SingletonSynchronized {

    private static volatile SingletonSynchronized instance;
    private final String _data;

    private SingletonSynchronized(String value) {
        _data = value;
    }

    public static SingletonSynchronized getInstance(String value) {
        if (instance == null) {
            synchronized (SingletonSynchronized.class) {
                if (instance == null) {
                    instance = new SingletonSynchronized(value);
                }
            }
        }
        return instance;
    }

    public String getData() {
        return _data;
    }
}

File: kotlin-sample/src/main/kotlin/patterns/creational/singleton/SingletonSynchronized.kt
package patterns.creational.singleton

import kotlin.concurrent.Volatile

class SingletonSynchronized private constructor(private val data: String) {

    companion object {
        @Volatile
        private var instance: SingletonSynchronized? = null

        fun getInstance(value: String): SingletonSynchronized? {
            if (instance == null) {
                synchronized(this) {
                    if (instance == null) {
                        instance = SingletonSynchronized(value)
                    }
                }
            }
            return instance
        }

        //Add this method for testing purposes
        fun reset() {
            synchronized(this) {
                instance = null
            }
        }
    }

    fun getData(): String = data
}
